rfcs/41-ui-navigation.md

```diff
@@ -0,0 +1,872 @@
+so the impact of this change is nil.
+
+## Future possibilities
+
+### New and better mouse picking!
+
+Since the focus algorithm does not handle input, it is not relevant
+to how we chose which entity is focused.
+
+The input methods can be interchanged without having to change or update
+the navigation plugins.
+
+### Multiple cursors
+
+A non-so-uncommon pattern of video games (especially split-screen)
+is to have two different players handle two different cursors on screen.
+
+The current design assumes a unique cursor navigating the menus
+and can't account for multi-cursor.
+However, it seems not fundamentally impossible to add multiple cursors.
+
+It would require modifying the `NavRequest` to include a sort of ID
+to identify which cursor originated the request,
+and keeping track in the `FocusState` which cursor the `Focused`, `Activated`
+and `Prioritized` variants are for.
+
+### Higher level wrappers
+
+In this RFC we discussed the possibility of improved API in a few places:
+- The `bevy-ui-navigation` [`event_helpers`] module, to simplify interacting
+  with `NavEvent`s
+- A visual wrapper for menu navigation that handles `display` `Style`s of menus
+  when they are entered and left
+
+## Drawbacks and design limitations
+
+### Major differences with the current focus implementation
+
+#### `Hovered` is not covered
+
+Since navigation is completely decoupled from input and ui lib,
+it is impossible for it to tell whether a focusable is hovered.
+To keep the hovering state functionality,
+it will be necessary to add it as an independent component
+separate from the navigation library.
+
+A fully generic navigation API cannot handle hover state.
+This doesn't mean we have to entirely give up hovering.
+It is simply added back as a component independent
+from the focus system.
+
+Note that the new `Hover` component does not comply with `FocusPolicy`.
+This allows reducing the complexity of the focus code.
+
+#### `Clicked` is not a focus state
+
+Compared to the current `Interaction` system, the new focus system
+separates semantics of "action" events and "UI focus state".
+
+Meaning that `Interaction::Clicked`'s equivalent in the new navigation system
+is the `NavEvent::NoChanges` event.
+
+This is important, as code relying on the "Clicked" focus state
+will need to be replace with reaction to a `NavEvent`.
+
+Looking at the code changes in the existing UI examples in [this RFC's PR]
+is a good indicator at how the code will need to be changed.
+
+### How does this work with the spawn/despawn workflow?
+
+The current design imposes to the user that all UI nodes such a buttons and
+menus must be already loaded, even if not visible on screen.
+This is to support re-focusing the last focused element when moving between
+menus.
+The design of this RFC _hinges on_ the whole tree been loaded in ECS while
+executing navigation requests, and would work poorly if menus were spawned and
+despawned dynamically.
```

Comment:
My mind keeps coming back to this statement. It is an extremely common use-case to not know the data ahead of time: I want to select a villager from a list. The game queries the list of villagers currently available, and spawns buttons for them so that I can click on one. Or just a search bar in an application.

Does this mean that we cannot support those use cases at all?

Reply:
Agreed, now that you highlight this I share your concerns.

Reply:
So. To  be precise, this section is really meant to contrast with the pattern of spawning and despawning each menu as they are created.

The limitation here (I've not tested it) I think is only when despawning the parent menu (for example despawning the main menu when entering the options menu), not when creating dynamically new menus you can enter, or despawning menus that do not have focus.

When you go back one level of menus (pressing backspace or clicking a "back" button), the navigation algo reads the `Entity` of the button the menu is "reachable from" (aka the button that activated the menu), and if that `Entity` is despawned, probably bad things happen (panic! most likely), as we now sort of "dereference a dangling `Entity`".

However, the reverse case of creating new menus at runtime is widely supported, the `MenuBuilder::ParentEntity` enum variant should help with runtime spawning of new menus, since you can just give it whatever `Entity` with a `Focusable` component you have available. It does mean however that the submenus must be created before the button is activated (for example when it gets focused/hovered).

It is not necessary for the whole tree to be loaded at once, but rather than the path back to the root menu is. For support for giving back focus to the last focused element (prioritized) you would also need to keep the `Focusable`s and menu tree of already-explored menus loaded. But typically, for a dropdown, this isn't something people expect.

This is why I mentioned an "infinitely deep recursive menu" as an edge case ui-navigation couldn't support.

I'm currently making a sort of "infinite weapon upgrade menu" as a cool example to demonstrate how easy it is to use your own navigation methods (using `Sprite` and `Text2d`) and generating new menus on the fly.

Reply:
When it comes to creating new `Focusable`s in an existing menu, it should be alright. But it's always going to be difficult when removing them (beside the mandatory minimum 1 `Focusable` in a menu). `TreeMenu` has a `active_child` field for avoiding iterating all children, but it completely violates the single-source-of-truth principle which makes despawning `Focusable`s dangerous (you must not despawn a focused or active focusable)

I'll probably have to rework access to the `active_child` to fallback to the ECS hierarchy when it doesn't exist anymore.

Reply:
That is good to know, and still impressive work with all of this. Is the last remaining hurdle then to find a good way to remove _leaf_ nodes in the navigation tree dynamically?

The requirement that the path back to root must exist I think is acceptable as long as we can dynamically remove leaf nodes